anxcloud/resource_dns_record.go

```diff
@@ -0,0 +1,225 @@
+package anxcloud
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	"go.anx.io/go-anxcloud/pkg/utils/object/compare"
+)
+
+var syncDNSRecordOps sync.Mutex
+var dnsRecordSkipMutexLock = providerContextKey("dns-record-skip-mutex-lock")
+
+func resourceDNSRecord() *schema.Resource {
+	return &schema.Resource{
+		Description:   "This resource allows you to create DNS records for a specified zone. Records of type TXT might be unstable for now.",
+		CreateContext: resourceDNSRecordCreate,
+		ReadContext:   resourceDNSRecordRead,
+		UpdateContext: resourceDNSRecordUpdate,
+		DeleteContext: resourceDNSRecordDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(time.Minute),
+			Read:   schema.DefaultTimeout(time.Minute),
+			Update: schema.DefaultTimeout(time.Minute),
+			Delete: schema.DefaultTimeout(time.Minute),
+		},
+		Schema: schemaDNSRecord(),
+	}
+}
+
+func resourceDNSRecordCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	defer lockRecordContextIfNeeded(ctx)()
+
+	a := m.(providerContext).api
```

Comment:
we have this line of code around a lot, maybe we should have a single `apiFromProviderConfig()` function to use everywhere, in case we ever change where the generic client comes from